<img src="https://github.com/cristiandarioortegayubro/BDS/blob/main/images/Logo%20Statsmodels.png?raw=true" width="280">
</p>

# **<font color="#07a8ed">Estadística inferencial con Python**


## **<font color="#07a8ed">Bibliotecas**

### **<font color="#07a8ed">Para análisis de datos**

In [1]:
import numpy as np
import pandas as pd

### **<font color="#07a8ed">Para gráficos**

In [2]:
import plotly.express as px
import plotly.graph_objects as go

### **<font color="#07a8ed">Para modelado**

#### **<font color="#07a8ed">Con Scikit-learn**

In [3]:
from sklearn.linear_model import LinearRegression
from sklearn import metrics
from sklearn.feature_selection import RFE

#### **<font color="#07a8ed">Con Statsmodels**

In [4]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

## **<font color="#07a8ed">Conjunto de Datos**

In [5]:
url = "https://raw.githubusercontent.com/cristiandarioortegayubro/BA/main/Datasets/publicidad_multiple.csv"

In [6]:
datos = pd.read_csv(url, index_col = 0)

In [7]:
datos.head(10)

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
1    44.5   39.3  45.1       10.4
2    17.2   45.9  69.3        9.3
3   151.5   41.3  58.5       18.5
4   180.8   10.8  58.4       12.9
5     8.7   48.9  75.0        7.2
6    57.5   32.8  23.5       11.8
7   120.2   19.6  11.6       13.2
8     8.6    2.1   1.0        4.8
9   199.8    2.6  21.2       10.6

<p align="justify">
Supóngase que el departamento de ventas de una empresa quiere estudiar la influencia que tiene la publicidad a través de distintos canales sobre el número de ventas de un producto. Se dispone de un conjunto de datos que contiene los ingresos (en millones) conseguido por ventas en 200 regiones, así como la cantidad de presupuesto, también en millones, destinado a anuncios por radio, TV y periódicos en cada una de ellas.


In [8]:
X = datos[["radio","tv","periodico"]]
y = datos[["ventas"]]

### **<font color="#07a8ed">Gráfico de distribución para cada variable**

In [9]:
variables = datos.columns.tolist()
variables

['ventas', 'radio', 'tv', 'periodico']

In [10]:
for i in variables:
  fig = px.histogram(datos,
                     x = i,
                     template = "gridon",
                     marginal= "box",
                     title = i).update_layout(bargap=0.2)
  fig.show()

### **<font color="#07a8ed">Scatterplot de cada variable explicativa**

In [11]:
for i in X.columns:
  fig = px.scatter(datos,
             x = i,
             y = y.ventas,
             template = "gridon",
             title = i)
  fig.show()

## **<font color="#07a8ed">MODELO 1: todas las variables y todos los datos**

**<font color="#07a8ed">Con statsmodels formula:**

In [14]:
datos.head(2)

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
1    44.5   39.3  45.1       10.4

In [12]:
modelo_f = smf.ols(formula="ventas~radio+tv+periodico", data=datos)
modelo_f = modelo_f.fit()

In [13]:
print(modelo_f.summary())

                            OLS Regression Results                            
Dep. Variable:                 ventas   R-squared:                       0.847
Model:                            OLS   Adj. R-squared:                  0.844
Method:                 Least Squares   F-statistic:                     360.8
Date:                Sun, 29 Sep 2024   Prob (F-statistic):           1.61e-79
Time:                        15:10:32   Log-Likelihood:                -986.30
No. Observations:                 200   AIC:                             1981.
Df Residuals:                     196   BIC:                             1994.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -33.2885      7.172     -4.641      0.0

<p align="justify">
El modelo con todas las variables introducidas como predictores tiene un $𝑅^2$  alto (0.847), es capaz de explicar el 84.7% de la variabilidad observada en las ventas.
<br><br>
F estadísitco grande (F-statistic =	360.8) y p-valor asociado extremadamente pequeño (Prob (F-statistic) = 1.61e-79), el modelo en su conjunto es estadísticamente significativo.
<br><br>
La variable explicativa tv no es estadísticamente significativa (p-valor = 0.703), por lo cual armaremos un nuevo modelo (<b>Modelo 2</b>) que no la incluya.

In [15]:
datos.corr()  ### analizamos las correlaciones

ventas     radio        tv  periodico
ventas     1.000000  0.054809  0.056648   0.782224
radio      0.054809  1.000000  0.354104   0.576223
tv         0.056648  0.354104  1.000000   0.228299
periodico  0.782224  0.576223  0.228299   1.000000

**<font color="#07a8ed">Con statsmodels api:**

In [16]:
X.head(2)

radio    tv  periodico
0   37.8  69.2       22.1
1   39.3  45.1       10.4

In [17]:
X = sm.add_constant(X, prepend=True)
modelo_api = sm.OLS(endog = y, exog = X,)
modelo_api = modelo_api.fit()
print(modelo_api.summary())

                            OLS Regression Results                            
Dep. Variable:                 ventas   R-squared:                       0.847
Model:                            OLS   Adj. R-squared:                  0.844
Method:                 Least Squares   F-statistic:                     360.8
Date:                Sun, 29 Sep 2024   Prob (F-statistic):           1.61e-79
Time:                        15:54:29   Log-Likelihood:                -986.30
No. Observations:                 200   AIC:                             1981.
Df Residuals:                     196   BIC:                             1994.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -33.2885      7.172     -4.641      0.0

**<font color="#07a8ed">Con scikit-learn:**

[Documentación](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)

In [18]:
X = datos[["radio","tv","periodico"]]

In [19]:
modelo_sk = LinearRegression()
modelo_sk.fit(X, y)
print(modelo_sk.intercept_)
print(modelo_sk.coef_)

[-33.28853795]
[[-3.44959312  0.04503377 18.48503551]]


### **<font color="#07a8ed">Interpretación**

In [20]:
datos.columns

Index(['ventas', 'radio', 'tv', 'periodico'], dtype='object')

In [21]:
modelo_f.params.round(2)

Intercept   -33.29
radio        -3.45
tv            0.05
periodico    18.49
dtype: float64

$$ventas = -\ 33.29 - 3.45\ radio + 0.05\ tv + 18.49\ periodico$$

In [22]:
print(f"{datos.columns[0]} = {round(modelo_f.params[0],2)} {round(modelo_f.params[1],2)}{datos.columns[1]} + {round(modelo_f.params[2],2)}{datos.columns[2]} + {round(modelo_f.params[3],2)}{datos.columns[3]}")

ventas = -33.29 -3.45radio + 0.05tv + 18.49periodico


<ipython-input-22-e5c1a51c7ffe>:1: FutureWarning:

Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`



### **<font color="#07a8ed">Predicción y evaluación del modelo**

In [23]:
X.head(2)

radio    tv  periodico
0   37.8  69.2       22.1
1   39.3  45.1       10.4

In [24]:
prediccion_1 = modelo_f.predict(X)

In [25]:
tabla = pd.DataFrame({"Prediccion":round(prediccion_1,1),
                      "Real": y.ventas,
                      "Ventas_medias": np.mean(y.ventas)})
tabla.head()

Prediccion   Real  Ventas_medias
0       248.0  230.1       147.0425
1        25.4   44.5       147.0425
2       -16.6   17.2       147.0425
3       168.9  151.5       147.0425
4       170.5  180.8       147.0425

**<font color="#07a8ed">Standardised mean squared error (SMSE):**

$$SMSE = \sqrt{MSE}=\sqrt{\frac{1}{n}\sum_{i=1}^{n}(y_i-\hat{y})^2}$$

In [26]:
SMSE = np.sqrt(sum((tabla.Real-tabla.Prediccion)**2)/(len(tabla))).round(2)
SMSE

33.54

**<font color="#07a8ed">Mean Absolute Error(MAE):**

$$MAE=\frac{1}{n}\sum_{i=1}^{n}|y_i-\hat{y}|$$

In [27]:
MAE = round(sum(abs(tabla.Real-tabla.Prediccion))/(len(tabla)),2)
MAE

25.66

$R_2:$

$$R_2 = 1-\frac{\frac{1}{n}\sum_{i=1}^{n}(y_i-\hat{y})^2}{\frac{1}{n}\sum_{i=1}^{n}(y_i-\bar{y})^2}$$

Donde:
* $\hat{y}$ es la predicción de la variable respuesta,
* $\bar{y}$ es el promedio de la variable respuesta.

In [28]:
R2= round(1-((sum((tabla.Real-tabla.Prediccion)**2)/(len(tabla)))/(sum((tabla.Real-tabla.Ventas_medias)**2)/(len(tabla)))),2)
R2

0.85

In [29]:
datos.periodico.iloc[0]

22.1

In [30]:
#Ejemplo 1 variable (R2-gráfico)
fig = px.scatter(datos,
                 x="periodico",
                 y="ventas",
                 trendline="ols",
                 trendline_color_override="darkorange",
                 template="gridon"
)

fig.add_shape(type="line",
              x0=datos.periodico.min(), y0=np.mean(y.ventas), x1=datos.periodico.max(), y1=np.mean(y.ventas),
              line=dict(color="Red",width=2)
)

Podemos calcular las métricas anteriores (calculadas manualmente) usando el módulo `metrics` de `sklearn`:

In [31]:
SMSE = round(metrics.mean_squared_error(y, prediccion_1,squared=False),2) #Error cuadrático medio
MAE = round(metrics.mean_absolute_error(y, prediccion_1),2) #Error absoluto medio
R2 = round(metrics.r2_score(y, prediccion_1),4)
print("SMSE: {}".format(SMSE))
print("MAE: {}".format(MAE))
print("R2: {}".format(R2))

SMSE: 33.53
MAE: 25.66
R2: 0.8467


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning:

'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.



In [32]:
predicciones = modelo_f.get_prediction(X).summary_frame(alpha=0.1).round(2)
predicciones.head()

mean  mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  obs_ci_upper
0  247.95     5.69         238.54         257.36        191.18        304.72
1   25.42     5.31          16.65          34.19        -31.25         82.08
2  -16.59     7.60         -29.16          -4.03        -73.97         40.78
3  168.85     4.38         161.61         176.09        112.40        225.30
4  170.54     5.16         162.02         179.07        113.91        227.17

### **<font color="#07a8ed">Multicolinealidad**

In [33]:
corr_matrix = round(X.corr(),3)

In [34]:
px.imshow(corr_matrix,
          title = "Matriz de correlacion",
          text_auto = True,
          template = "gridon",
          labels={"color":"Coeficiente"})

Las variables predictoras radio y periodico está correlacionadas. Es recomendable seleccionar una de ellas.

## **<font color="#07a8ed">MODELO 2: periódico y radio y todos los datos**

In [35]:
modelo_2 = smf.ols(formula="ventas~radio+periodico", data=datos).fit()
print(modelo_2.summary())

                            OLS Regression Results                            
Dep. Variable:                 ventas   R-squared:                       0.847
Model:                            OLS   Adj. R-squared:                  0.845
Method:                 Least Squares   F-statistic:                     543.4
Date:                Sun, 29 Sep 2024   Prob (F-statistic):           6.56e-81
Time:                        16:04:09   Log-Likelihood:                -986.38
No. Observations:                 200   AIC:                             1979.
Df Residuals:                     197   BIC:                             1989.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -32.5203      6.869     -4.734      0.0

<p align="justify">
Los p-valores son extremadamente pequeño, garantizando que las variables son estadísticamente significativas y por lo tanto los parámetros no son 0.

### **<font color="#07a8ed">Interpretación**

$$ventas = -32.52 - 3.43\ radio + 18.49\ periodico$$

### **<font color="#07a8ed">Predicción y evaluación del modelo**

In [36]:
X.head(2)

radio    tv  periodico
0   37.8  69.2       22.1
1   39.3  45.1       10.4

In [37]:
prediccion_2 = modelo_2.predict(X)

In [38]:
tabla = pd.DataFrame({"Prediccion":round(prediccion_2,1),
                      "Real":y.ventas})
tabla.head(8)

Prediccion   Real
0       246.6  230.1
1        25.1   44.5
2       -17.9   17.2
3       168.0  151.5
4       169.0  180.8
5       -67.0    8.7
6        73.3   57.5
7       144.4  120.2

In [39]:
SMSE = round(metrics.mean_squared_error(y, prediccion_2, squared=False),2) #Error cuadrático medio
MAE = round(metrics.mean_absolute_error(y, prediccion_2),2) #Error absoluto medio
R2 = round(metrics.r2_score(y, prediccion_2),4)
print("SMSE: {}".format(SMSE))
print("MAE: {}".format(MAE))
print("R2: {}".format(R2))

SMSE: 33.55
MAE: 25.63
R2: 0.8466


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning:

'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.



## **<font color="#07a8ed">MODELO 3: periódico y tv y todos los datos**

In [40]:
datos.head(3)

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
1    44.5   39.3  45.1       10.4
2    17.2   45.9  69.3        9.3

In [41]:
modelo_3 = smf.ols(formula="ventas~tv+periodico", data=datos).fit()
print(modelo_3.summary())

                            OLS Regression Results                            
Dep. Variable:                 ventas   R-squared:                       0.628
Model:                            OLS   Adj. R-squared:                  0.624
Method:                 Least Squares   F-statistic:                     166.0
Date:                Sun, 29 Sep 2024   Prob (F-statistic):           5.61e-43
Time:                        16:04:41   Log-Likelihood:                -1075.0
No. Observations:                 200   AIC:                             2156.
Df Residuals:                     197   BIC:                             2166.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -24.7325     11.121     -2.224      0.0

Este modelo disminuye considerablemente el $𝑅^2$, no se tendrá en cuenta.

In [42]:
prediccion_3 = modelo_3.predict(X)
SMSE = round(metrics.mean_squared_error(y, prediccion_3, squared=False),2) #Error cuadrático medio
MAE = round(metrics.mean_absolute_error(y, prediccion_3),2) #Error absoluto medio
R2 = round(metrics.r2_score(y, prediccion_3),4)
print("SMSE: {}".format(SMSE))
print("MAE: {}".format(MAE))
print("R2: {}".format(R2))

SMSE: 52.26
MAE: 41.11
R2: 0.6276


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning:

'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.



## **<font color="#07a8ed">MODELO 4: periódico y radio y resultado positivo por región**

In [43]:
datos.head()

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
1    44.5   39.3  45.1       10.4
2    17.2   45.9  69.3        9.3
3   151.5   41.3  58.5       18.5
4   180.8   10.8  58.4       12.9

El dataframe contiene por cada región, las ventas y la inversión en los canales de publicidad.

In [44]:
datos.sum()

ventas       29408.5
radio         4652.8
tv            6110.8
periodico     2804.5
dtype: float64

Se puede observar que las ventas totales superan ampliamente los gastos en canales de publicidad.

In [45]:
datos["utilidad"] = datos.ventas - datos.radio - datos.tv - datos.periodico

In [46]:
datos.head()

ventas  radio    tv  periodico  utilidad
0   230.1   37.8  69.2       22.1     101.0
1    44.5   39.3  45.1       10.4     -50.3
2    17.2   45.9  69.3        9.3    -107.3
3   151.5   41.3  58.5       18.5      33.2
4   180.8   10.8  58.4       12.9      98.7

In [47]:
utilidad = round(datos.utilidad.sum(),2)
print(f"La utilidad total es: {utilidad}")

La utilidad total es: 15840.4


### **<font color="#07a8ed">Filtrando por utilidad positiva**

In [48]:
utilidad = datos.query("utilidad > 0")
utilidad = utilidad.reset_index(drop = True)
utilidad.drop(columns = "utilidad", inplace = True)
utilidad

ventas  radio    tv  periodico
0     230.1   37.8  69.2       22.1
1     151.5   41.3  58.5       18.5
2     180.8   10.8  58.4       12.9
3     120.2   19.6  11.6       13.2
4       8.6    2.1   1.0        4.8
..      ...    ...   ...        ...
148    38.2    3.7  13.8        7.6
149    94.2    4.9   8.1        9.7
150   177.0    9.3   6.4       12.8
151   283.6   42.0  66.2       25.5
152   232.1    8.6   8.7       13.4

[153 rows x 4 columns]

In [49]:
X_utilidad = utilidad[['radio', 'tv', 'periodico']]
y_utilidad = utilidad["ventas"]

In [50]:
for i in X_utilidad.columns:
  fig = px.scatter(utilidad,
             x = i,
             y = y_utilidad,
             template = "gridon",
             title = i)
  fig.show()

### **<font color="#07a8ed">Creación del modelo**

In [51]:
utilidad

ventas  radio    tv  periodico
0     230.1   37.8  69.2       22.1
1     151.5   41.3  58.5       18.5
2     180.8   10.8  58.4       12.9
3     120.2   19.6  11.6       13.2
4       8.6    2.1   1.0        4.8
..      ...    ...   ...        ...
148    38.2    3.7  13.8        7.6
149    94.2    4.9   8.1        9.7
150   177.0    9.3   6.4       12.8
151   283.6   42.0  66.2       25.5
152   232.1    8.6   8.7       13.4

[153 rows x 4 columns]

In [52]:
modelo_4 = smf.ols(formula="ventas~radio+periodico", data=utilidad).fit()
print(modelo_4.summary())

                            OLS Regression Results                            
Dep. Variable:                 ventas   R-squared:                       0.863
Model:                            OLS   Adj. R-squared:                  0.861
Method:                 Least Squares   F-statistic:                     473.0
Date:                Sun, 29 Sep 2024   Prob (F-statistic):           1.67e-65
Time:                        16:07:13   Log-Likelihood:                -711.31
No. Observations:                 153   AIC:                             1429.
Df Residuals:                     150   BIC:                             1438.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -107.7656      9.595    -11.232      0.0

### **<font color="#07a8ed">Interpretación**

$$ventas = -\ 107.77 - 6.68\ radio + 27.80\ periodico$$

### **<font color="#07a8ed">Predicción y evaluación del modelo**

In [53]:
X_utilidad.head(2)

radio    tv  periodico
0   37.8  69.2       22.1
1   41.3  58.5       18.5

In [54]:
prediccion_4 = modelo_4.predict(X_utilidad)
prediccion_4

0      253.876450
1      130.391224
2      178.647704
3      128.146184
4       11.635642
          ...    
148     78.778918
149    129.136374
150    185.897477
151    320.315054
152    207.258447
Length: 153, dtype: float64

In [55]:
SMSE = round(metrics.mean_squared_error(y_utilidad, prediccion_4, squared=False),2) #Error cuadrático medio
MAE = round(metrics.mean_absolute_error(y_utilidad, prediccion_4),2) #Error absoluto medio
R2 = round(metrics.r2_score(y_utilidad, prediccion_4),4)
print("SMSE: {}".format(SMSE))
print("MAE: {}".format(MAE))
print("R2: {}".format(R2))

SMSE: 25.28
MAE: 19.64
R2: 0.8631


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_regression.py:492: FutureWarning:

'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.



In [56]:
px.scatter_3d(utilidad,
              x="radio",
              y="periodico",
              z="ventas")

##**Comparación de modelos**

| Modelo | SMSE | MAE | R2 |
|--|------|------|-------|
|1 |33.53 | 25.66| 0.8467|
|2 |33.55 | 25.63| 0.8466|
|3 |52.26 | 41.11| 0.6276|
|4 |**25.28** | **19.64**| **0.8631**|